File: src/cosmic_pi_network/agi_framework/nlp_interface.py
import spacy

class NLPInterface:
    def __init__(self):
        self.nlp = spacy.load("en_core_web_sm")

    def process_input(self, input_data):
        doc = self.nlp(input_data)
        self.entities = [ent.text for ent in doc.ents]

    def get_entities(self):
        return self.entities

    def __str__(self):
        return f"NLPInterface(entities={self.entities})"

    def __repr__(self):
        return repr(self.entities)

File: src/cosmic_pi_network/agi_framework/cpp/NLPInterface.cpp
#include "NLPInterface.h"
#include <spacy/spacy.h>

NLPInterface::NLPInterface() {
    this->nlp = new spacy::Language("en_core_web_sm");
}

void NLPInterface::process_input(const char* input_data) {
    spacy::Doc doc = nlp->process(input_data);
    int num_entities = doc.ents.size();
    this->entities = new char*[num_entities];
    for (int i = 0; i < num_entities; i++) {
        entities[i] = doc.ents[i].text.c_str();
    }
}

const char** NLPInterface::get_entities() {
    return entities;
}

NLPInterface::~NLPInterface() {
    delete nlp;
    delete[] entities;
}
